### ISSUE TITLE
How to export virtual counters from monitoring module

### ISSUE THREAD

**Issue by faziano92** [jcsaezal/pmctrack#10]
I have written a monitoring module which periodically reads LLC misses and cycles, and computes  a first metric, i.e. rate. I also have a second metric which is only the value of LLC misses counted. I can successfully use these values in my custom scheduler, which is great ! Then I also wrote a multi-threaded user-space application which uses libpmctrack to count the LLC misses, as in the example in test/test_libpmctrack/basic-test-libpmctrack. If the dummy monitoring module is active I can read some values, however if I activate my monitoring module by echoing the respective index in `/proc/pmc/mm_manager` I do not read any value and the libpmctrack functions failed in trying to read `/proc/pmc/config file`. This should happen because I haven't exported the counters as virtual counters. Looking at the dummy monitoring module and at the ipc_sampling module, I added a callback `mymodule_counter_usage()`, which sets the following fields in the usage structure:

-  `usage->hwpmc_mask = 0x0;`
- `usage->nr_virtual_counters = 1;`
- `usage->nr_experiments = 1;`
- `usage->vcounter_desc[0] = "LLC miss count"`

Also I added the following lines in `mymodule_on_new_sample()` callback to update the virtual counters:
`
if ((prof->virt_counter_mask & 0x1)) {
		sample->virt_mask |= 0x1;
		sample->nr_virt_counts++;
		sample->virtual_counts[0] = llc_data->llc_misses;
}
`

In this way I can get some values, but I am not sure what I am actually reading as well as I would like to know more on how to properly export the counters and therefore be able to gather counter values from user applications with libpmctrack that are running on a system which uses at the same time my monitoring module in its scheduler.

Thank you in advance for you help and for pmctrack !

**Comment by jcsaezal** [comment]
Hi @faziano92,

It looks like you are exporting virtual counters correctly. I would avoid spaces in the name of the virtual counter, though ("llc_miss_count" seems a better choice).

In that scenario, the monitoring module does use performance counters. So, you must explicitly indicate which physical counters you are using in the `usage->hwpmc_mask` bitmask. That is the correct way to do it, to avoid issues. Moreover, in using HW counters from the module, you cannot set specific PMC configurations from user-space. For example, if you use the `pmctrack` command-line tool you are not allowed to use "-c <pmc_configuration>" (that generates the error you mentioned) but only indicate which virtual counters you want to retrieve with -V. The PMC configuration is forced by the active monitoring module, but the counter values will appear anyway. The behavior when using libpmctrack directly is similar to that.

In summary, what you are doing (using hardware counters inside a monitoring module) only makes sense, if you want to communicate the value of a PMC "high-level" metric to the OS scheduler (in the kernel). However, if you want user-space programs to be able to alter the configuration of PMCs, you must disable that monitoring module.

I hope that helps!
